src/Control/Monad/Aff.js

```diff
@@ -762,26 +762,33 @@ var Aff = function () {
         case APPLY:
           lhs = head._1._3;
           rhs = head._2._3;
-          // We can only proceed if both sides have resolved.
-          if (lhs === EMPTY || rhs === EMPTY) {
-            return;
-          }
-          // If either side resolve with an error, we should continue with
-          // the first error.
-          if (util.isLeft(lhs)) {
-            if (util.isLeft(rhs)) {
-              if (fail === lhs) {
-                fail = rhs;
-              }
-            } else {
-              fail = lhs;
-            }
-            step    = null;
-            head._3 = fail;
-          } else if (util.isLeft(rhs)) {
-            step    = null;
-            fail    = rhs;
+          // If we have a failure we should kill the other side because we
+          // can't possible yield a result anymore.
+          if (fail) {
             head._3 = fail;
+            tmp     = true;
+            kid     = killId++;
+
+            kills[kid] = kill(early, fail === lhs ? head._2 : head._1, function (/* unused */) {
```

Comment:
This looks tricky. Don't understand the encoding well enough to say if this is correct.

Reply:
* `head` = the current Applicative node in the stack
* `head._1` = the left-hand-side node
* `head._2` = the right-hand-side node
* `fail` = the resolved exception
* `lhs` = the resolved value of the left-hand-side parallel operation (`head._1._3`)
* `rhs` = the resolved value of the right-hand-side parallel operation (`head._2._3`)

So, if the left-hand-side resolved to a failure, kill the right-hand-side, otherwise kill the left-hand-side.

Reply:
I don't think I actually need `rhs` anymore, so I can lose that.

Reply:
Nevermind, it's used in the success branches.